runtime/interpreter/value.go

```diff
@@ -19443,26 +19443,64 @@ func (v *SomeValue) Storable(
 	maxInlineSize uint64,
 ) (atree.Storable, error) {
 
+	// SomeStorable returned from this function can be encoded in two ways:
+	// - if innermost value is too large, innermost value is encoded in a separate slab
+	//   while SomeStorable wrapper is encoded inline with reference to slab containing
+	//   innermost value.
+	// - otherwise, SomeStorable with innermost value is encoded inline.
+	// The above applies to both immutable innermost value (such as StringValue),
+	// and mutable innermost value (such as ArrayValue).
+
 	if v.valueStorable == nil {
-		var err error
-		v.valueStorable, err = v.value.Storable(
+
+		innermostValue, nestedLevels := v.getInnermostValue()
+
+		someStorableEncodedPrefixSize := getSomeStorableEncodedPrefixSize(nestedLevels)
+
+		// Reduce maxInlineSize for innermost value to make sure
+		// that SomeStorable wrapper is always encoded inline.
+		maxInlineSize -= uint64(someStorableEncodedPrefixSize)
+
+		innermostValueStorable, err := innermostValue.Storable(
 			storage,
 			address,
 			maxInlineSize,
 		)
 		if err != nil {
 			return nil, err
 		}
+
+		valueStorable := innermostValueStorable
+		for i := 1; i < int(nestedLevels); i++ {
+			valueStorable = SomeStorable{
+				Storable: valueStorable,
+			}
+		}
+		v.valueStorable = valueStorable
 	}
 
-	return maybeLargeImmutableStorable(
-		SomeStorable{
-			Storable: v.valueStorable,
-		},
-		storage,
-		address,
-		maxInlineSize,
-	)
+	// No need to call maybeLargeImmutableStorable() here for SomeStorable because:
+	// - encoded SomeStorable size = someStorableEncodedPrefixSize + innermost storable size
+	// - innermost storable size < maxInlineSize - someStorableEncodedPrefixSize
+	return SomeStorable{
+		Storable: v.valueStorable,
+	}, nil
+}
+
+// getInnermostValue returns innermost value of SomeValue and nested levels.
+// For example, SomeValue{true} has innermost value true, and nested levels 1.
+func (v *SomeValue) getInnermostValue() (atree.Value, uint64) {
```

Comment:
Maybe rename the function to indicate that this function only gets the innermost value _as long as there are `SomeValue`s_ / traverses `SomeValue`s, and for `Some(Some([Some(Some(Some(1)))])) (2 Some, Array, 3 Some), the array, and not the integer `1`